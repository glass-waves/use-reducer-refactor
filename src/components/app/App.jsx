/* eslint-disable max-len */
import React, { useReducer } from 'react';

/* 
## Test

Write a test that tests the behavior off App.

## Refactor

Refactor the hook to `useReducer` instead of `useState`. Refactor only the hook, you
shouldn't need to refactor the component at all.

## Rubric

- behavior test 3 pts
- refactor 7 pts



*/

const useRecord = (init) => {
  const reducer = (state, action) => {
    switch (action.type) {
      case 'RECORD':
        return {
          ...state,
          before: [...state.before, state.current],
          current: action.payload,
        };
      case 'UNDO':
        return {
          before: state.before.slice(0, -1),
          current: state.before[state.before.length - 1],
          after: [state.current, ...state.after],
        };
      case 'REDO':
        return {
          before: [...state.before, state.current],
          current: state.after[0],
          after: state.after.slice(1),
        };
      default:
        return state;
    }
  };
  const initialState = {
    before: [],
    current: init,
    after: [],
  };
  const [state, dispatch] = useReducer(reducer, initialState);


  const undo = () => {
    dispatch({ type: 'UNDO' });
  };

  const redo = () => {
    dispatch({ type: 'REDO' });
  };

  const record = (val) => {
    dispatch({ type: 'RECORD', payload: val });
  };
  const { current } = state;
  return {
    undo,
    record,
    redo,
    current,
  };
};

function App() {
  const { current, undo, redo, record } = useRecord('#FF0000');

  return (
    <>
      <button onClick={undo} aria-label="undo">
        undo
      </button>
      <button onClick={redo} aria-label="redo">
        redo
      </button>
      <input
        data-testid="colorpicker"
        type="color"
        value={current}
        onChange={({ target }) => record(target.value)}
      />
      <div
        data-testid="colordiv"
        style={{ backgroundColor: current, width: '10rem', height: '10rem' }}
      ></div>
    </>
  );
}

export default App;
