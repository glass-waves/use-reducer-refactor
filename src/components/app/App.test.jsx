import React from 'react';
import { render, cleanup, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

describe('App component', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(<App />);
    const undo = screen.getByRole('button', { name: 'undo' });
    const redo = screen.getByRole('button', { name: 'redo' });
    const div = screen.getByTestId('colordiv');
    const colorPicker = screen.getByTestId('colorpicker');
    expect(div).toHaveStyle('background-color: #FF0000');
    fireEvent.change(colorPicker, { target: { value: '#FFFFFF' } });
    expect(div).toHaveStyle('background-color: #FFFFFF');
    userEvent.click(undo);
    expect(div).toHaveStyle('background-color: #FF0000');
    userEvent.click(redo);
    expect(div).toHaveStyle('background-color: #FFFFFF');
    
    expect(asFragment()).toMatchSnapshot();
  });
});
