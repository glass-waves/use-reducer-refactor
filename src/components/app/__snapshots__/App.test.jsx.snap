// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App 1`] = `
<DocumentFragment>
  <button
    aria-label="undo"
  >
    undo
  </button>
  <button
    aria-label="redo"
  >
    redo
  </button>
  <input
    data-testid="colorpicker"
    type="color"
    value="#ffffff"
  />
  <div
    data-testid="colordiv"
    style="background-color: rgb(255, 255, 255); width: 10rem; height: 10rem;"
  />
</DocumentFragment>
`;
